.github/workflows/matrix.yml

```diff
@@ -239,23 +253,27 @@ jobs:
       with:
         path: ${{ steps.get-vars.outputs.ccache-path }}
         # double-dash after compiler is not a typo, it is to disambiguate between g++-<date> and g++-11-<date> for restore key prefix matching
-        key: ccache-${{ runner.os }}-${{ matrix.compiler }}--${{ steps.get-vars.outputs.date }}
+        key: ccache-${{ github.ref_name }}-${{ runner.os }}-${{ matrix.compiler }}--${{ steps.get-vars.outputs.datetime }}
         restore-keys: |
-          ccache-${{ runner.os }}-${{ matrix.compiler }}--
+          ccache-master-${{ runner.os }}-${{ matrix.compiler }}--
```

Comment:
This will force PRs, when they miss, which they will, to always try to fetch the latest master artifact. There's no race condition, there's a unidirectional flow of artifacts from master to PRs. PRs then drop the artifacts on the floor and on next run re-restore from master, possibly the same artifacts, possibly newer ones.